### ISSUE TITLE
 EC policy not supported

### ISSUE THREAD

**Issue by hert525** [colinmarc/hdfs#226]
proto: required field "dirList.partialListing.ecPolicy.cellSize" not set

**Comment by colinmarc** [comment]
Hi @hert525,

Thanks for the bug report! Can you provide more information about how to reproduce this error?

**Comment by mnotti** [comment]
@colinmarc getting same error:
`proto: required field "fs.ecPolicy.cellSize" not set and stack unwound: goroutine 883`

Happens when reading from a dir with EC on hadoop3 hdfs cluster
go/pkg/mod/github.com/colinmarc/hdfs/v2@v2.1.1/readdir.go:10 ReadDir method is what fails for me when attempting to read any dir that is erasure coded or any parent of that dir. ReadDir will work just fine if reading other dirs.

**Comment by mrtvfuencxozd** [comment]
Hi,
having this issue here with a hadoop 3.3.0 namenode, I looked at [hdfs.proto](https://github.com/apache/hadoop/blob/release-3.3.0-RC0/hadoop-hdfs-project/hadoop-hdfs-client/src/main/proto/hdfs.proto) file from hadoop repo and found that definition of `ErasureCodingPolicyProto` differs.

I will test more, but modifying `internal/protocol/hadoop_hdfs/hdfs.proto` as below's diff, I can now list files.
```
@@ -332,12 +332,21 @@ message ECSchemaProto {
   required uint32 parityUnits = 3;
   repeated ECSchemaOptionEntryProto options = 4;
 }
+/**
+ * EC policy state.
+ */
+enum ErasureCodingPolicyState {
+  DISABLED = 1;
+  ENABLED = 2;
+  REMOVED = 3;
+}

 message ErasureCodingPolicyProto {
-  required string name = 1;
-  required ECSchemaProto schema = 2;
-  required uint32 cellSize = 3;
+  optional string name = 1;
+  optional ECSchemaProto schema = 2;
+  optional uint32 cellSize = 3;
   required uint32 id = 4; // Actually a byte - only 8 bits used
+  optional ErasureCodingPolicyState state = 5 [default = ENABLED];
 }
```

**Comment by zhanglistar** [comment]
> Hi,
> having this issue here with a hadoop 3.3.0 namenode, I looked at [hdfs.proto](https://github.com/apache/hadoop/blob/release-3.3.0-RC0/hadoop-hdfs-project/hadoop-hdfs-client/src/main/proto/hdfs.proto) file from hadoop repo and found that definition of `ErasureCodingPolicyProto` differs.
> 
> I will test more, but modifying `internal/protocol/hadoop_hdfs/hdfs.proto` as below's diff, I can now list files.
> 
> ```
> @@ -332,12 +332,21 @@ message ECSchemaProto {
>    required uint32 parityUnits = 3;
>    repeated ECSchemaOptionEntryProto options = 4;
>  }
> +/**
> + * EC policy state.
> + */
> +enum ErasureCodingPolicyState {
> +  DISABLED = 1;
> +  ENABLED = 2;
> +  REMOVED = 3;
> +}
> 
>  message ErasureCodingPolicyProto {
> -  required string name = 1;
> -  required ECSchemaProto schema = 2;
> -  required uint32 cellSize = 3;
> +  optional string name = 1;
> +  optional ECSchemaProto schema = 2;
> +  optional uint32 cellSize = 3;
>    required uint32 id = 4; // Actually a byte - only 8 bits used
> +  optional ErasureCodingPolicyState state = 5 [default = ENABLED];
>  }
> ```

@colinmarc  does this right?

**Comment by mrtvfuencxozd** [comment]
Hi an extra comment,
I've used this change for a while (using rclone as client app).
It is fine to browse/read files, however it seems that some other changes might be required for proper write support.

I can write files, but they end in a strange state (not cleanly closed from what namenode logs suggest) where they use much more space than required (and cause lot of errors in namenode logs)

**Comment by zhanglistar** [comment]
> having this issue here with a hadoop 3.3.0 namenode, I looked at [hdfs.proto](https://github.com/apache/hadoop/blob/release-3.3.0-RC0/hadoop-hdfs-project/hadoop-hdfs-client/src/main/proto/hdfs.proto) file from hadoop repo and found that definition of `ErasureCodingPolicyProto` differs.

How to compile it? I have tried to do the patch, but got many compile errors.

**Comment by mrtvfuencxozd** [comment]
I also had some errors, probably caused by protobuf compiler/version.

As at that time, I did not find pointers to a clean way of builging `.proto` files, I ended up using an ubuntu 20.04 container to build the just content of `internal/protocol/hadoop_hdfs` and copied generated .go files to my dev tree.

I don't remember the exact commands though.

**Comment by colinmarc** [comment]
I think this should be fixed in c67abc15ff977efec63cd84ea31aad4fa55439ac. Please tell me if you encounter it again!
